### PR TITLE
Prevent potential dispatch of HttpTunnel events to a dead consumer.

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
Before you ask, yes, this was causing crashes in production.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
@SolidWallOfCode So maybe this is a 7.1.0 candidate?

File: proxy/http/HttpTunnel.h [hunk 1]
```diff
@@ -475,12 +475,28 @@ HttpTunnel::get_producer(HttpTunnelType_t type)
 inline HttpTunnelConsumer *
 HttpTunnel::get_consumer(VConnection *vc)
 {
-  for (int i = 0; i < MAX_CONSUMERS; i++) {
-    if (consumers[i].vc == vc) {
-      return consumers + i;
+  /** Rare but persistent problem in which a @c INKVConnInternal is used a consumer, released, and
```
Comment (zwoop) [inline proxy/http/HttpTunnel.h]: Nitpick as shit, but likely a grammatical typo here? "...as a consumer..." maybe"

File: proxy/http/HttpTunnel.h [hunk 2]
```diff
@@ -475,12 +475,28 @@ HttpTunnel::get_producer(HttpTunnelType_t type)
...
+      It's not correct to check for the consumer being alive - at a minimum `HTTP_TUNNEL_EVENT_DONE`
+      is dispatched against a consumer after the consumer is not alive. Instead if a non-alive
+      consumer matches it is stored as a candidate and returned if no other match is found. If a
+      live matching consumer is found, it is immediately returned. It is never valid to have the
+      same VC in more than one active consumer. This should avoid a performance impact because in
+      the usual case the consumer will be alive.
+  */
+  HttpTunnelConsumer *zret = nullptr;
+  for (HttpTunnelConsumer &c : consumers) {
+    if (c.vc == vc) {
```
Comment (zwoop) [inline proxy/http/HttpTunnel.h]: Hmmm, so the logic here is not the same as the old code. As far as I can tell, the old code returned the "next" consumer, whereas this code returns the one that matches? The latter logic seems more reasonable, so curious why the old worked at all ?

Comment (SolidWallOfCode) [inline proxy/http/HttpTunnel.h]: The "next" consumer and "the one that matches" are the same thing in almost all cases. In general every "vc" (which are `InkContInternal` instances) has a distinct address so there is only one match, which is by definition therefore the "next" consumer. In both the new and old code, the consumer vc is checked for a pointer match against the argument vc. The logic differs only in the case the same vc address is in the table for two consumers.

The old code worked fine as long as every `InkContInternal` allocated during the span of a transaction had a distinct address, which is quite normally the case. Although the free list is a FIFO, if any other transaction or thread allocated between a transaction's release and realloc there would not be a problem. You need just the right amount of load (some, but not too much) and the right timing on the transactions on the same `ET_NET` thread to trigger this.

File: proxy/http/HttpTunnel.h (hunk 1 above)
Comment (SolidWallOfCode) [inline proxy/http/HttpTunnel.h]: As you will, PHB. I need to add a bit more comment anyway.

**Comment by masaori335** [comment]
I changed milestone and label. Also I added this to 7.x releases Project. Because it looks like this is not backported to 7.1.x branch.
https://github.com/apache/trafficserver/commits/7.1.x/proxy/http/HttpTunnel.h

**Comment by zwoop** [comment]
Cheery-picked to 7.1.x

**Pull request merged**